src/cppgc.rs

```diff
@@ -714,3 +725,73 @@ impl<T: GarbageCollected + std::fmt::Display> std::fmt::Display for Ptr<T> {
     std::fmt::Display::fmt(&**self, fmt)
   }
 }
+
+/// `GcCell` is a zero-overhead memory cell that provides interior mutability
+/// for garbage collected types.
+///
+/// Mutable access to the value inside the `GcCell` is granted by delivering
+/// proof that the caller has mutable access to the V8 Isolate (such as through
+/// a [`HandleScope`](crate::HandleScope)). This statically guarantees that
+/// access patterns follow the rules of both the garbage collector and the Rust
+/// borrow checker.
+///
+/// `GcCell` also implements [`Traced`], which means that it can hold other
+/// garbage collected references, like [`Member`] or [`WeakMember`]. If the
+/// `GcCell` holds other garbage collected objects, those objects cannot be
+/// accessed while the `GcCell` is borrowed, and the caller must construct
+/// temporary pointers ([`Ptr<T>`]) on the stack in order to access nested
+/// objects in the object graph.
+pub struct GcCell<T: ?Sized> {
+  // Contents guarded by access to the `Isolate`.
+  value: UnsafeCell<T>,
+}
+
+unsafe impl<T: Send + ?Sized> Send for GcCell<T> {}
+unsafe impl<T: Sync + ?Sized> Sync for GcCell<T> {}
+
+impl<T> GcCell<T> {
+  pub fn new(value: T) -> Self {
+    Self {
+      value: UnsafeCell::new(value),
+    }
+  }
+
+  pub fn set(&self, isolate: &mut crate::Isolate, value: T) {
+    _ = isolate;
```

Comment:
For soundness I think you need a check that the isolate in question matches the isolate that owns `self`.

Reply:
Yeah, this is tricky, since objects in practice do not move between threads. A runtime check perhaps defeats the purpose.

If it is actually possible for two separate isolates to get access to the same pointer in safe code, I would be more interested in an API change, such as making `new()` unsafe, to document valid use.

The important thing is to not need unsafe on every access.